Adding motif finding tutorial using the stats.meta.stackexchange.com data dump

rjurney opened the pull request:
This PR makes the following additions to create a tutorial on motif finding using [`stats.meta.stackexchange.com` data dump at the internet archive](https://archive.org/details/stackexchange). Teaching the concepts behind this powerful tool will drive increased adoption of GraphFrames.

* A new tutorial on motif finding in `docs/motif-tutorial.md`
* Code for the demo in `python/graphframes/examples` - `download.py`, `xml_to_parquet.py`, `graph.py` and `motif.py`
* A new data folder `python/graphframes/examples/data`

This code was originally written by myself under the [MIT License](https://github.com/Graphlet-AI/graphml-class/blob/main/LICENSE) for a class at [Connected Data London 2024](https://2024.connected-data.london) called [Full Stack Graph Machine Learning](https://2024.connected-data.london/talks/full-stack-graph-machine-learning/). It can be found at https://github.com/Graphlet-AI/graphml-class/tree/main/graphml_class/stats

bjornjorgensen commented:
too big PR :) 

can we use notebooks for this? something like https://spark.apache.org/docs/3.5.4/api/python/getting_started/quickstart_ps.html

rjurney commented:
> too big PR :)
> 
> can we use notebooks for this? something like https://spark.apache.org/docs/3.5.4/api/python/getting_started/quickstart_ps.html

We could use notebooks, but if the content isn't integrated with the other documentation, a lot fewer people will see and use it. The actual motif tutorial isn't many lines of code - the edits to the README and index page are what are so long.

Some of this reworks the main docs / README and some of it deals with a motif tutorial. Lemme see if I can split out.

rjurney commented:
As mentioned in #481 I want to use this to expand the unit tests to use a larger dataset of millions of edges.

rjurney commented:
@SauronShepherd @bjornjorgensen Can you please review this PR? My motif finding tutorial is finally ready :) I want to ship it and then cut a new release. It includes a new extended README and other improvements. Please forgive me for the size - it got out of hand - I'll create smaller PRs in the future.

SemyonSinchenko commented:
@rjurney Can we split this PR to series of smaller PRs? At least separate infrastructure part (CI, build, gitignore, etc.) and tutorial itself?

SauronShepherd commented:
I agree on that, Sem.

I've only reviewed the first ones, but I have some doubts:

- The only differences in many lines seem to be the end character. Is that
ok?
https://github.com/graphframes/graphframes/pull/473/commits/13c5e74eda1918e26242e74fb7bfc7d2a155c2a8

- Why mentioning explictly a concrete IDE in the .gitignore? Maybe that's
something every developed should do on its own according to their IDE.
.vscode
https://github.com/graphframes/graphframes/pull/473/commits/13194348c967c211516d202fc5125f0f9ef8cb9a

- Why excluding a data folder that maybe shouldn't be located inside the
project in the first place? python/graphframes/examples/data
Why download local test data inside the project?
https://github.com/graphframes/graphframes/pull/473/commits/8d84baaab44f7f467b58e553f6c4aabddfef1a18

These are not critical points, they just crossed my mind while having a
look to the PR.

El sáb, 8 feb 2025 a las 12:31, Sem ***@***.***>) escribió:

> @rjurney <https://github.com/rjurney> Can we split this PR to series of
> smaller PRs? At least separate infrastructure part (CI, build, gitignore,
> etc.) and tutorial itself?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/graphframes/graphframes/pull/473#issuecomment-2645092234>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACCN674CC7WXYV3VP4G3ZHT2OXTJDAVCNFSM6AAAAABUFSGKESVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMNBVGA4TEMRTGQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

rjurney commented:
Offhand I didn't know how to break it up, but I will figure it out and do so.

SemyonSinchenko commented:
> Offhand I didn't know how to break it up, but I will figure it out and do so.

The most important changes that blocks other PRs are related to the `setup.py`, `requirements` and changes in python tests. Can you separate these changes from the tutorial itself and the downloading scripts?

bjornjorgensen commented on python/graphframes/tests.py:
```diff
@@ -15,63 +15,72 @@
-from pyspark import SparkContext
-from pyspark.sql import functions as sqlfunctions, SparkSession
+import pytest
+from pyspark import SparkConf, SparkContext
```
make one PR where you update tests.. 
are you sure that we should be using pytest and not unittests? 

Apache spark are using unittest.

bjornjorgensen commented on python/graphframes/tests.py:
```diff
@@ -15,63 +15,72 @@
             raise TypeError("invalid minor version")
         try:
-            version_info['minor'] = int(m.group(2))
+            version_info["minor"] = int(m.group(2))
```
I guess this was for python2 and python3 versions.

bjornjorgensen commented on python/graphframes/tests.py:
```diff
@@ -15,63 +15,72 @@
+        cls.conf = SparkConf().setAppName("GraphFramesTests")
+        cls.conf.set(
+            "spark.submit.pyFiles",
+            os.path.abspath("python/dist/graphframes-{VERSION}-py3-none-any.whl"),
```
what is this for?

bjornjorgensen commented on python/graphframes/tests.py:
```diff
@@ -123,28 +135,38 @@ def setUp(self):
+
+        assert gtu.spark_at_least_of_version("1.7")
+        assert gtu.spark_at_least_of_version("2.0")
+        assert gtu.spark_at_least_of_version("2.0.1")
```
What is this for?

bjornjorgensen commented on python/graphframes/tests.py:
```diff
@@ -281,83 +322,79 @@ def test_aggregate_messages(self):
         assert len(all1) == 1
         labels2 = labels.filter("id >= 5").select("label").collect()
-        all2 = set([x.label for x in labels2])
+        all2 = {row.label for row in labels2}
```
what is this? 
change a set to dict?

rjurney commented:
@SemyonSinchenko @bjornjorgensen thank you guys VERY much for these reviews! Would you recommend I split it up before addressing the issues, or address the issues before splitting it up into multiple PRs?

SemyonSinchenko commented:
> @SemyonSinchenko @bjornjorgensen thank you guys VERY much for these reviews! Would you recommend I split it up before addressing the issues, or address the issues before splitting it up into multiple PRs?

I would recommend to leave it as is for now and open a series of small PRs, related to CI, pytest, build, etc.

rjurney commented:
Okay guys, diving into splitting this PR up...

rjurney commented:
@SauronShepherd @SemyonSinchenko @bjornjorgensen please have a look at #511 - the actual documentation portion of the PR. I will do a second and third one now for the docs code and build improvement stuff.

rjurney commented:
@SauronShepherd @SemyonSinchenko @bjornjorgensen @WeichenXu123 okay also created #512 and #513. I want to try to merge these and ship a new release this coming week, in advance of the GraphFrames Hackathon.

rjurney commented:
Closing in favor of #511, #512 and #518.
